fairseq2n/src/fairseq2n/data/sample_data_source.cc

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "fairseq2n/data/sample_data_source.h"
+
+#include <ATen/CPUGeneratorImpl.h>
+#include <ATen/Context.h>
+#include <ATen/Functions.h>
+
+#include <stdexcept>
+
+#include "fairseq2n/utils/tensor.h"
+
+namespace fairseq2n::detail {
+
+sample_data_source::sample_data_source(std::vector<data_pipeline> &&pipelines, std::vector<float32> &&weights)
+    : pipelines_(std::move(pipelines))
+{
+    weights_ = make_tensor_from_vector(weights, { static_cast<std::int64_t>(pipelines_.size()) });
+    generator_ = at::globalContext().defaultGenerator(c10::DeviceType::CPU);
+}
+
+std::optional<data>
+sample_data_source::next()
+{
+    if (eod_)
+        return std::nullopt;
+
+    std::optional<data> output = pipelines_[next_index()].next();
+    if (!output)
+        eod_ = true;
+
+    return output;
+}
+
+void
+sample_data_source::reset()
+{
+    generator_ = at::globalContext().defaultGenerator(c10::DeviceType::CPU);
```

Comment:
`defaultGenerator` won't change throughout the lifetime of the process. You can move this call to constructor.

Reply:
True, I missed that. All generator state handling will be done outside of this operator